### PR TITLE
Feature/teams-addmultisite-for-user

### PR DESCRIPTION
**Pull request by KoR82**

### CODE REVIEW

File: backend/app/Http/Requests/Api/GetBounceRateChartHttpRequest.php [hunk 1]
```diff
@@ -5,9 +5,54 @@
...
 use App\Http\Request\ApiFormRequest;
-use App\Http\Requests\ChartHttpRequestTrait;
+use App\Rules\Timestamp;
+use App\Rules\TimestampAfter;
+use App\Rules\IsWebsiteRelatedWithUser;
 
 final class GetBounceRateChartHttpRequest extends ApiFormRequest
 {
-    use ChartHttpRequestTrait;
+    public function rules(): array
```
[the remark this answers is not in the document]
Comment (KoR82) [reply backend/app/Http/Requests/Api/GetBounceRateChartHttpRequest.php]: fixed

File: backend/app/Http/Requests/Visit/GetPageViewsChartAvgTimeHttpRequest.php [hunk 2]
```diff
@@ -5,9 +5,54 @@
...
 
 use App\Http\Request\ApiFormRequest;
-use App\Http\Requests\ChartHttpRequestTrait;
+use App\Rules\Timestamp;
+use App\Rules\TimestampAfter;
+use App\Rules\IsWebsiteRelatedWithUser;
 
 final class GetPageViewsChartAvgTimeHttpRequest extends ApiFormRequest
 {
-    use ChartHttpRequestTrait;
```
[the remark this answers is not in the document]
Comment (KoR82) [reply backend/app/Http/Requests/Visit/GetPageViewsChartAvgTimeHttpRequest.php]: fixed

File: backend/app/Http/Requests/Visit/GetPageViewsFilterHttpRequest.php [hunk 3]
```diff
@@ -5,9 +5,53 @@
...
 
 use App\Http\Request\ApiFormRequest;
-use App\Http\Requests\ChartHttpRequestTrait;
+use App\Rules\Timestamp;
+use App\Rules\TimestampAfter;
+use App\Rules\IsWebsiteRelatedWithUser;
 
 final class GetPageViewsFilterHttpRequest extends ApiFormRequest
 {
-    use ChartHttpRequestTrait;
```
[the remark this answers is not in the document]
Comment (KoR82) [reply backend/app/Http/Requests/Visit/GetPageViewsFilterHttpRequest.php]: fixed

File: backend/app/Http/Requests/Visit/GetUniquePageViewsChartHttpRequest.php [hunk 4]
```diff
@@ -4,9 +4,54 @@
...
 
 use App\Http\Request\ApiFormRequest;
-use App\Http\Requests\ChartHttpRequestTrait;
+use App\Rules\Timestamp;
+use App\Rules\TimestampAfter;
+use App\Rules\IsWebsiteRelatedWithUser;
 
 class GetUniquePageViewsChartHttpRequest extends ApiFormRequest
 {
-    use ChartHttpRequestTrait;
```
[the remark this answers is not in the document]
Comment (KoR82) [reply backend/app/Http/Requests/Visit/GetUniquePageViewsChartHttpRequest.php]: fixed

File: backend/app/Http/Requests/Visitor/GetNewChartVisitorsHttpRequest.php [hunk 5]
```diff
@@ -5,9 +5,54 @@
...
 
 use App\Http\Request\ApiFormRequest;
-use App\Http\Requests\ChartHttpRequestTrait;
+use App\Rules\Timestamp;
+use App\Rules\TimestampAfter;
+use App\Rules\IsWebsiteRelatedWithUser;
 
 final class GetNewChartVisitorsHttpRequest extends ApiFormRequest
 {
-    use ChartHttpRequestTrait;
```
[the remark this answers is not in the document]
Comment (KoR82) [reply backend/app/Http/Requests/Visitor/GetNewChartVisitorsHttpRequest.php]: fixed

File: backend/app/Http/Requests/Visitor/GetVisitorsBounceRateHttpRequest.php [hunk 6]
```diff
@@ -5,9 +5,53 @@
 namespace App\Http\Requests\Visitor;
 
 use App\Http\Request\ApiFormRequest;
-use App\Http\Requests\ChartHttpRequestTrait;
+use App\Rules\IsWebsiteRelatedWithUser;
 
 final class GetVisitorsBounceRateHttpRequest extends ApiFormRequest
 {
-    use ChartHttpRequestTrait;
```
[the remark this answers is not in the document]
Comment (KoR82) [reply backend/app/Http/Requests/Visitor/GetVisitorsBounceRateHttpRequest.php]: fixed

[the remark this answers is not in the document]
Comment (KoR82) [reply backend/app/Http/Requests/Visitor/GetVisitorsBounceRateHttpRequest.php]: fixed

File: backend/app/Http/Requests/Visitor/GetChartTotalVisitorsByDateRangeHttpRequest.php [hunk 7]
```diff
@@ -5,9 +5,54 @@
...
 
 use App\Http\Request\ApiFormRequest;
-use App\Http\Requests\ChartHttpRequestTrait;
+use App\Rules\Timestamp;
+use App\Rules\TimestampAfter;
+use App\Rules\IsWebsiteRelatedWithUser;
 
 final class GetChartTotalVisitorsByDateRangeHttpRequest extends ApiFormRequest
 {
-    use ChartHttpRequestTrait;
```
[the remark this answers is not in the document]
Comment (KoR82) [reply backend/app/Http/Requests/Visitor/GetChartTotalVisitorsByDateRangeHttpRequest.php]: fixed

File: backend/app/Http/Requests/Session/GetSessionsFilterHttpRequest.php [hunk 8]
```diff
@@ -5,9 +5,53 @@
...
 
 use App\Http\Request\ApiFormRequest;
-use App\Http\Requests\ChartHttpRequestTrait;
+use App\Rules\Timestamp;
+use App\Rules\TimestampAfter;
+use App\Rules\IsWebsiteRelatedWithUser;
 
 final class GetSessionsFilterHttpRequest extends ApiFormRequest
 {
-    use ChartHttpRequestTrait;
```
[the remark this answers is not in the document]
Comment (KoR82) [reply backend/app/Http/Requests/Session/GetSessionsFilterHttpRequest.php]: fixed

File: backend/app/Repositories/EloquentWebsiteRepository.php [hunk 9]
```diff
@@ -32,4 +32,40 @@ public function getByTrackNumber(int $trackNumber): ?Website
...
+            $firstTeamMemberWebsite = $userWebsites->filter(function($website) {
+                return $website->pivot->role == 'member';
+            })->first();
+        }
+
+        if ($firstTeamMemberWebsite) {
+            return $firstTeamMemberWebsite;
+        } else {
+            throw new WebsiteNotFoundException;
+        }
```
[the remark this answers is not in the document]
Comment (KoR82) [reply backend/app/Repositories/EloquentWebsiteRepository.php]: fixed

File: backend/app/Repositories/EloquentWebsiteRepository.php [hunk 10]
```diff
@@ -32,4 +32,40 @@ public function getByTrackNumber(int $trackNumber): ?Website
...
+            return $website->pivot->role == 'owner';
+        })->first();
+
+        if ($firstOwnWebsite) {
+            return $firstOwnWebsite;
+        } else {
+            $firstTeamMemberWebsite = $userWebsites->filter(function($website) {
+                return $website->pivot->role == 'member';
+            })->first();
+        }
```
[the remark this answers is not in the document]
Comment (KoR82) [reply backend/app/Repositories/EloquentWebsiteRepository.php]: fixed

File: backend/app/Actions/Visitors/GetAllActivityVisitorRequest.php [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    public static function fromRequest(GetAllActivityVisitorHttpRequest $request): self
+    {
+        return new static($request->websiteId());
+    }
+
+    public function websiteId(): int
+    {
+        return $this->websiteId;
+    }
+}
```
[the remark this answers is not in the document]
Comment (KoR82) [reply backend/app/Actions/Visitors/GetAllActivityVisitorRequest.php]: This request with 1 parameter = website_id

[the remark this answers is not in the document]
Comment (mes1901) [reply backend/app/Actions/Visitors/GetAllActivityVisitorRequest.php]: before the changes it was empty

File: backend/app/Actions/Visitors/GetNewVisitorsAction.php [hunk 12]
```diff
@@ -15,9 +15,12 @@ public function __construct(VisitorRepository $visitorRepository)
         $this->visitorRepository = $visitorRepository;
     }
 
-    public function execute(): GetNewVisitorsResponse
+    public function execute(GetAllActivityVisitorRequest $request): GetNewVisitorsResponse
     {
-        $visitors = $this->visitorRepository->newest();
+        $visitors = $this->visitorRepository->newest()
+                       ->filter(function($visitor) use ($request) {
+                           return $visitor->website_id == $request->websiteId();
```
Comment (lenchvolodymyr) [inline backend/app/Actions/Visitors/GetNewVisitorsAction.php]: strict comparison

File: backend/app/Repositories/EloquentWebsiteRepository.php [hunk 13]
```diff
@@ -32,4 +32,40 @@ public function getByTrackNumber(int $trackNumber): ?Website
...
+
+        throw new WebsiteNotFoundException;
+    }
+
+    public function getFirstExistingUserWebsite(): Website
+    {
+        $userWebsites = auth()->user()->websites;
+
+        $firstOwnWebsite = $userWebsites->filter(function($website) {
+            return $website->pivot->role == 'owner';
```
Comment (lenchvolodymyr) [inline backend/app/Repositories/EloquentWebsiteRepository.php]: strict comparison

File: backend/app/Repositories/EloquentWebsiteRepository.php [hunk 14]
```diff
@@ -32,4 +32,40 @@ public function getByTrackNumber(int $trackNumber): ?Website
             throw new WebsiteNotFoundException;
         }
     }
+
+    public function getCurrentWebsite(int $websiteId): Website
+    {
+        $userWebsitesIds = auth()->user()
```
Comment (lenchvolodymyr) [inline backend/app/Repositories/EloquentWebsiteRepository.php]: use repository to retrive websites by userId

File: backend/app/Rules/IsWebsiteRelatedWithUser.php [hunk 15]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+namespace App\Rules;
+
+use Illuminate\Contracts\Validation\Rule;
+
+class IsWebsiteRelatedWithUser implements Rule
```
Comment (lenchvolodymyr) [inline backend/app/Rules/IsWebsiteRelatedWithUser.php]: grammatically preposition "to" is used "related to"

File: backend/app/Entities/User.php [hunk 16]
```diff
@@ -43,16 +43,23 @@ public function sendPasswordResetNotification($token)
...
     {
-        return $this->hasOne(Website::class, 'user_id');
+        return $this->belongsToMany(Website::class)->withPivot('role');
     }
 
-    public function getWebsite()
+    public function isWebsiteOwner(int $websiteId)
     {
-        return $this->website;
+        $userWebsite = auth()->user()
```
Comment (lenchvolodymyr) [inline backend/app/Entities/User.php]: why do you retrieve in model auth user? Check on the current user, or better all query to DB do inside the repository

Comment (lenchvolodymyr) [reply backend/app/Entities/User.php]: rename method to hasWebsite(int $websiteId), and check on current user, not authorized. If you need you can call it as auth()->user()->hasWebsite($websiteId), but not inside the model

File: backend/app/Actions/Website/AddWebsiteAction.php [hunk 17]
```diff
@@ -26,11 +26,14 @@ public function execute(AddWebsiteRequest $request): AddWebsiteResponse
         $website->name = $request->getName();
         $website->domain = $request->getDomain();
         $website->single_page = $request->getSinglePage();
-        $website->user_id = Auth::id();
         $website->tracking_number = $this->getLastTrackingNumber() + 1;
 
         $this->websiteRepository->save($website);
 
+        auth()->user()->websites()->attach($website->id, [
```
Comment (lenchvolodymyr) [inline backend/app/Actions/Website/AddWebsiteAction.php]: attach role via repository

**Pull request merged**